integration-testing/src/main/java/io/grpc/testing/integration/AbstractTransportTest.java

```diff
@@ -409,72 +407,57 @@ public void halfDuplexCallShouldSucceed() throws Exception {
+        StreamingOutputCallResponse.newBuilder()
+            .setPayload(Payload.newBuilder()
+                .setType(PayloadType.COMPRESSABLE)
+                .setBody(ByteString.copyFrom(new byte[100001]))).build());
 
-    // Build the request.
-    List<Integer> responseSizes = Arrays.asList(50, 100, 150, 200);
-    StreamingOutputCallRequest.Builder streamingOutputBuilder =
-        StreamingOutputCallRequest.newBuilder();
-    streamingOutputBuilder.setResponseType(COMPRESSABLE);
-    for (Integer size : responseSizes) {
-      streamingOutputBuilder.addResponseParametersBuilder().setSize(size).setIntervalUs(0);
-    }
-    StreamingOutputCallRequest request = streamingOutputBuilder.build();
+    long start = System.nanoTime();
 
-    // Start the call and prepare capture of results.
-    final List<StreamingOutputCallResponse> results =
-        Collections.synchronizedList(new ArrayList<StreamingOutputCallResponse>());
-    final SettableFuture<Void> completionFuture = SettableFuture.create();
-    final AtomicInteger count = new AtomicInteger();
+    final ArrayBlockingQueue<Object> queue = new ArrayBlockingQueue<Object>(10);
+    Call<StreamingOutputCallRequest, StreamingOutputCallResponse> call =
+        channel.newCall(TestServiceGrpc.CONFIG.streamingOutputCall);
     call.start(new Call.Listener<StreamingOutputCallResponse>() {
-
       @Override
-      public void onHeaders(Metadata.Headers headers) {
-      }
+      public void onHeaders(Metadata.Headers headers) {}
 
       @Override
       public void onPayload(final StreamingOutputCallResponse payload) {
-        results.add(payload);
-        count.incrementAndGet();
+        queue.add(payload);
       }
 
       @Override
       public void onClose(Status status, Metadata.Trailers trailers) {
-        if (status.isOk()) {
-          completionFuture.set(null);
-        } else {
-          completionFuture.setException(status.asException());
-        }
+        queue.add(status);
       }
     }, new Metadata.Headers());
-
-    // Send the request.
     call.sendPayload(request);
     call.halfClose();
 
-    // Slowly set completion on all of the futures.
-    int expectedResults = responseSizes.size();
-    while (count.get() < expectedResults) {
-      // Allow one more inbound message to be delivered to the application.
-      call.request(1);
-
-      // Sleep a bit.
-      Uninterruptibles.sleepUninterruptibly(1, TimeUnit.SECONDS);
-    }
+    // Time how long it takes to get the first response.
+    call.request(1);
+    Object o = queue.poll(1, TimeUnit.SECONDS);
+    long duration = System.nanoTime() - start;
+    assertEquals(goldenResponses.get(0), o);
 
-    // Wait for successful completion of the response.
-    completionFuture.get();
+    // Without giving additional flow control, make sure that we don't get another response.
+    assertNull(queue.poll(Math.max(duration * 4, 1000000), TimeUnit.NANOSECONDS));
```

Comment:
This is not supposed to complete with the specified time, because request(1) hasn't been called. So I just based how long to wait on 4*duration of getting the first request (with newCall overhead). Since that could have been very fast (say, with a in-process testing-only transport) I require the delay to be at least 1ms.

Added comment.